src/jit/hwintrinsicxarch.cpp

```diff
@@ -137,9 +140,18 @@ NamedIntrinsic Compiler::lookupHWIntrinsic(const char* methodName, InstructionSe
             if (isa == hwIntrinsicInfoArray[i].isa && strcmp(methodName, hwIntrinsicInfoArray[i].intrinsicName) == 0)
             {
                 result = hwIntrinsicInfoArray[i].intrinsicID;
+                break;
             }
         }
+
+        if ((NI_SW_INTRINSIC_START < result) && (result < NI_SW_INTRINSIC_END))
+        {
+            return NI_Illegal;
```

Comment:
Unnecessary.

Reply:
IMO having a new flag and going through intrinsic recognition deep up to the point where we return nullptr after checking flag creates longer and less efficient codepath. Can you explain what type of benefits will we have by implementing your proposal and what is difference in code path between tthis two solution and why your poposal is more efficient?

Reply:
> checking flag creates longer and less efficient codepath. 

I do not think there is any issue of efficiency. Clear code is much more important.

> explain what type of benefits will we have by implementing your proposal

1. Managed-implemented intrinsics also can use the code of throwing `PlatformNotSupported` or type `NotSupported`.
2. We will move certain implementations to CodGen in the future.
3. A new range `SW_INTRINSIC_START\SW_INTRINSIC_END` is not necessary.

In summary, I just want to change the code as less as possible.

Reply:
> flag creates longer and less efficient codepath.

As I said, if you want to cut off the intrinsic-recognition/importation earlier, I think just removing the definition of the managed-implemented helper from this table is okay.

Reply:
Code where we will be returning `NI_Illegal` or `nullptr` is inside `Compiler::impIntrinsic` in importer.cpp:

```C++
    if ((methodFlags & CORINFO_FLG_JIT_INTRINSIC) != 0)
    {
        // The recursive calls to Jit intrinsics are must-expand by convention.
        mustExpand = mustExpand || gtIsRecursiveCall(method);

        if (intrinsicID == CORINFO_INTRINSIC_Illegal)
        {
            ni = lookupNamedIntrinsic(method);

#ifdef FEATURE_HW_INTRINSICS
            if (ni > NI_HW_INTRINSIC_START && ni < NI_HW_INTRINSIC_END)
            {
                return impHWIntrinsic(ni, method, sig, mustExpand);
            }
#endif // FEATURE_HW_INTRINSICS
        }
    }
```

Reply:
If an intrinsic is not in the table, `lookupHWIntrinsic` will automatically return `NI_Illegal` for it. 
Do not need any change.

Reply:
We can modify `Compiler::impIntrinsic` to get even more efficient:

```C++
        if (intrinsicID == CORINFO_INTRINSIC_Illegal)
        {
            ni = lookupNamedIntrinsic(method);

#ifdef FEATURE_HW_INTRINSICS
            if (ni > NI_HW_INTRINSIC_START && ni < NI_SW_INTRINSIC_START)
            {
                return impHWIntrinsic(ni, method, sig, mustExpand);
            }
            else if (ni > NI_SW_INTRINSIC_START && ni < NI_SW_INTRINSIC_END)
            {
                return nullptr;
            }
#endif // FEATURE_HW_INTRINSICS
        }
```

Reply:
> We can modify Compiler::impIntrinsic to get even more efficient:

I do not think so.

Reply:
See example: https://github.com/dotnet/coreclr/compare/master...dotnetrt:SetAllVector128v3

Reply:
> > We can modify Compiler::impIntrinsic to get even more efficient:

> I do not think so.

Can you explain?

Reply:
> If an intrinsic is not in the table, lookupHWIntrinsic will automatically return NI_Illegal for it.
Do not need any change.

But than you would support any intrinsic which is not in the table without any way to assert it is correct to do so.

Reply:
If an intrinsic is not in the table, lookupHWIntrinsic will automatically return NI_Illegal for it. That is exactly same with what you are doing now, but no change required.

Reply:
> I do not think another group in the table is necessary, please just giving them a flag and checking the flag to return `nullptr` in `Compiler::impHWIntrinsic`.

Furthermore, how would you check the flag on intrinsic which is not in the table?

Reply:
To be clear, there are two solutions:
1. Remove the definition of managed-implemented intrinsic from the table, and no any other change required 
2. Checking flag to return nullptr, which needs the definition in the table.

Reply:
I would say that 2 is the better route overall, as it allows us to have explicit asserts/checks for these intrinsics elsewhere, if required.

@CarolEidt, do you have a preference?

Reply:
> To be clear, there are two solutions:

Why there are only 2 solutions available? Is it an objective, proven mathematically fact or it is just your opinion? If it's the second than it is up for discussion ...

Reply:
> I would say that 2 is the better route overall, as it allows us to have explicit asserts/checks for these intrinsics elsewhere, if required.

I am not a fan of flags usage since they are:

1. Not very well defined and not documented in a way which helps in using them.
2, Require additional check to use them.

Furthermore, the code path used in current solution simply treats managed implementation as ordinary managed function with `AggresiveInlining` attribute and inlines it.

There are no obvious needs for any further checks therefore implementing flag based solution for unclear future usage is an overengineering.

Reply:
IMO current 3rd solution allows to:

1. have positive asserts on all supported intrinsics.
2. saves 2 function calls
3. is more clear to follow overall for code users who have not created code (current preferences may be based on natural bias on using known, existing solution despite that it is more complex).

Reply:
I think we should simply remove these from the table and have it "just work". Anything else seems like over-engineering to me. I don't think any of the other alternatives are as clean. If they are in the table with a flag, IMO it pollutes the space of what are expected to be actual intrinsics (i.e. methods for which the *JIT* generates code directly). If they are in their own table, it begs the question of why they're there at all. They're not really intrinics.

If we find a need for something more complex down the road we can always add it.

Reply:
@CarolEidt Thanks for clear guidelines.